svm/src/account_loader.rs

```diff
@@ -95,6 +95,264 @@ pub struct FeesOnlyTransaction {
 }
 
+#[derive(Debug, Clone)]
+pub(crate) struct AccountLoader<'a, CB: TransactionProcessingCallback> {
+    pub program_cache: ProgramCacheForTxBatch,
+    account_overrides: Option<&'a AccountOverrides>,
+    account_cache: HashMap<Pubkey, AccountCacheItem>,
+    callbacks: &'a CB,
+}
+impl<'a, CB: TransactionProcessingCallback> AccountLoader<'a, CB> {
+    pub fn new(
+        callbacks: &'a CB,
+        account_overrides: Option<&'a AccountOverrides>,
+        program_cache: ProgramCacheForTxBatch,
+    ) -> AccountLoader<'a, CB> {
+        Self {
+            program_cache,
+            account_overrides,
+            account_cache: HashMap::new(),
+            callbacks,
+        }
+    }
+
+    pub fn load_account(
+        &mut self,
+        account_key: &Pubkey,
+        usage_pattern: AccountUsagePattern,
+    ) -> Option<LoadedTransactionAccount> {
+        let is_writable = usage_pattern.is_writable();
+        let is_invisible = usage_pattern.is_invisible();
+
+        if let Some(cache_item) = self.account_cache.get_mut(account_key) {
+            if !cache_item.inspected_as_writable {
+                // Inspecting a read-only account is harmless. Inspecting a writable
+                // account must be done before rent collection, and must never be done
+                // again in-batch, because inspection is intended to preserve the
+                // state prior to any writes.
+                self.callbacks.inspect_account(
+                    account_key,
+                    AccountState::Alive(&cache_item.account),
+                    is_writable,
+                );
+                cache_item.inspected_as_writable = is_writable;
+            }
+
+            // We must compare the full account, not just check if lamports are 0.
+            // Otherwise we might erroneously hide rent-delinquent read-only accounts.
+            if cache_item.account == AccountSharedData::default() {
+                None
+            } else if let Some(program) = is_invisible
+                .then_some(())
+                .and_then(|_| self.program_cache.find(account_key))
+            {
+                // NOTE if this account could be found in the program cache, we must
+                // substitute its size here, to perserve an oddity with transaction size
+                // calculations. However, we CANNOT skip the accounts cache check,
+                // in case the account was non-executable and modified in-batch.
+                // When we properly check if program cache entries are executable,
+                // we can go to the program cache first and remove this branch, because
+                // executable accounts are immutable.
+                Some(LoadedTransactionAccount {
+                    loaded_size: program.account_size,
+                    account: account_shared_data_from_program(&program),
+                    rent_collected: 0,
+                })
+            } else {
```

Comment:
this (checking the account cache first, but overriding with the program cache if appropriate) is the magic that makes the new loader behave exactly like the old loader for the program cache and transaction data size edge cases tested by #3045. it _also_ makes transactions that modify and then use programs in the same entry behave the same as if those transactions were in different entries in the same slot, a set of edge cases that are even _more_ subtle

to restate all known issues (except for loader size, which is irrelevant):
1. accounts from the program cache are assumed to be marked executable even if theyre not
2. transaction data size is different depending on whether an invoked program is an instruction account
3. when programs are deployed, they immediately become "executable" in the same entry despite a delayed visibility tombtone, because program cache tombstones are not checked by the account loader. they fail at execution because the bpf loader checks the slot. there is nothing to fix in this pr because this behavior is identical for programs deployed in the same slot in different entries
4. when programs are closed or resized, subsequent transactions that read programdata must see the changed state
5. when programs are closed or resized, subsequent transactions that invoke those programs must see the updated data size from the program cache. this is NOT related to simd83. nothing in account locking stops you from making a two-transaction entry that closes a program and then calls it: the first transaction takes a read lock on program and write lock on programdata, the second transaction takes a read lock on program. the transaction fails anyway, but the data size must be correct because whether it fails during loading or execution affects consensus

i verified by hand that 5 behaves the same with this pr, but it isnt feasible to test because it depends entirely on state internal to `load_and_execute_sanitized_transactions` because the per-batch program cache is (correctly) not merged back into the global cache. but tests that cover 1 and 2 are sufficient to demonstrate that the program cache is prioritized over the account cache for non-instruction accounts

as an aside, 1 and 5 cannot be fixed with a feature gate without a simd before fee-paying transactions are enabled, because there is no unambiguous "correct" behavior. as it stands, executable programs stay executable forever, but if we checked program cache tombstones we would treat them as non-executable. however, once fee-only transactions are enabled, we dont need a simd _or_ a feature gate, because both paths lead to the same result and execution becomes an implementation detail